packages/theme-language-server-common/src/formatting/providers/HtmlElementAutoclosingOnTypeFormattingProvider.ts

```diff
@@ -0,0 +1,113 @@
+import {
+  getName,
+  LiquidHTMLASTParsingError,
+  LiquidHtmlNode,
+  NodeTypes,
+  toLiquidHtmlAST,
+} from '@shopify/liquid-html-parser';
+import {
+  DocumentOnTypeFormattingParams,
+  Position,
+  Range,
+  TextEdit,
+} from 'vscode-languageserver-protocol';
+import { AugmentedSourceCode } from '../../documents';
+import { BaseOnTypeFormattingProvider } from '../types';
+import { findCurrentNode } from '../../visitor';
+import { TextDocument } from 'vscode-languageserver-textdocument';
+import { SetCursorPosition } from '../types';
+
+const defer = (fn: () => void) => setTimeout(fn, 10);
+
+/**
+ * This class is responsible for closing dangling HTML elements.
+ *
+ * Say user types <script>, then we'd want `</script>` to be inserted.
+ *
+ * Thing is we want to do that only if the `</script>` isn't already present in the file.
+ * If the user goes to edit `<script>` and types `>`, we don't want to insert `</script>` again.
+ *
+ * The "trick" we use here is to only add the `</script>` part if the
+ * document.ast is an instance of LiquidHTMLASTParsingError and that the
+ * unclosed element is of the correct name.
+ *
+ * @example:
+ * ```html
+ *   <div id="main">
+ *     <div id="inner">|
+ *   </div>
+ * ```
+ * - The user just finished typing `<div id="inner">` inside the div#main.
+ * - This parses as though the div#inner is closed and div#main isn't.
+ * - That's OK.
+ * - This makes a LiquidHTMLASTParsingError with unclosed div (the div#main).
+ * - Since
+ *     - the cursor is at the end of a div, and
+ *     - the unclosed element is a div,
+ *   Then we can insert one automatically after the cursor and fix the AST.
+ *
+ * ```html
+ *  <div id="main">
+ *    <div id="inner">|</div>
+ *  </div>
+ * ```
+ */
+export class HtmlElementAutoclosingOnTypeFormattingProvider
+  implements BaseOnTypeFormattingProvider
+{
+  constructor(private setCursorPosition: SetCursorPosition) {}
+
+  onTypeFormatting(
+    document: AugmentedSourceCode,
+    params: DocumentOnTypeFormattingParams,
+  ): TextEdit[] | null {
+    const textDocument = document.textDocument;
+    const ch = params.ch;
+    // position is position of cursor so 1 ahead of char
+    const { line, character } = params.position;
+    switch (ch) {
```

Comment:
I know this pattern is probably lifted from the other file but since there's only one case here I wonder if we should keep the `switch` at all? I guess it would make it easier to add more triggers later on?